dpctl/tensor/libtensor/source/where.cpp

```diff
@@ -0,0 +1,258 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file defines Python API for implementation functions of
+/// dpctl.tensor.where
+//===----------------------------------------------------------------------===//
+
+#include "dpctl4pybind11.hpp"
+#include <CL/sycl.hpp>
+#include <complex>
+#include <cstdint>
+#include <pybind11/complex.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <utility>
+
+#include "kernels/where.hpp"
+#include "simplify_iteration_space.hpp"
+#include "utils/memory_overlap.hpp"
+#include "utils/offset_utils.hpp"
+#include "utils/type_dispatch.hpp"
+#include "where.hpp"
+
+namespace dpctl
+{
+namespace tensor
+{
+namespace py_internal
+{
+
+namespace _ns = dpctl::tensor::detail;
+
+using dpctl::tensor::kernels::search::where_contig_impl_fn_ptr_t;
+using dpctl::tensor::kernels::search::where_strided_impl_fn_ptr_t;
+
+static where_contig_impl_fn_ptr_t where_contig_dispatch_table[_ns::num_types]
+                                                             [_ns::num_types];
+static where_strided_impl_fn_ptr_t where_strided_dispatch_table[_ns::num_types]
+                                                               [_ns::num_types];
+
+using dpctl::utils::keep_args_alive;
+
+std::pair<sycl::event, sycl::event>
+py_where(dpctl::tensor::usm_ndarray condition,
+         dpctl::tensor::usm_ndarray x1,
+         dpctl::tensor::usm_ndarray x2,
+         dpctl::tensor::usm_ndarray dst,
+         sycl::queue exec_q,
+         const std::vector<sycl::event> &depends)
+{
+
+    if (!dpctl::utils::queues_are_compatible(exec_q, {x1, x2, condition, dst}))
+    {
+        throw py::value_error(
+            "Execution queue is not compatible with allocation queues");
+    }
+
+    int nd = condition.get_ndim();
+    int x1_nd = x1.get_ndim();
+    int x2_nd = x2.get_ndim();
+    int dst_nd = dst.get_ndim();
+
+    if (nd != x1_nd || nd != x2_nd) {
+        throw py::value_error(
+            "Input arrays are not of appropriate dimension for where kernel.");
+    }
+
+    if (nd != dst_nd) {
+        throw py::value_error(
+            "Destination is not of appropriate dimension for where kernel.");
+    }
+
+    const py::ssize_t *x1_shape = x1.get_shape_raw();
+    const py::ssize_t *x2_shape = x2.get_shape_raw();
+    const py::ssize_t *dst_shape = dst.get_shape_raw();
+    const py::ssize_t *cond_shape = condition.get_shape_raw();
+
+    bool shapes_equal(true);
+    size_t nelems(1);
+    for (int i = 0; i < nd; ++i) {
+        nelems *= static_cast<size_t>(dst_shape[i]);
+        shapes_equal = shapes_equal && (x1_shape[i] == dst_shape[i]) &&
+                       (x2_shape[i] == dst_shape[i]) &&
+                       (cond_shape[i] == dst_shape[i]);
```

Comment:
Perhaps use a temp variable for `dst_shape[i]` for readability:

```suggestion
        const auto &sh_i = dst_shape[i];
        nelems *= static_cast<size_t>(sh_i);
        shapes_equal = shapes_equal && (x1_shape[i] == sh_i) &&
                       (x2_shape[i] == sh_i) &&
                       (cond_shape[i] == sh_i);
```